src/librustc_mir/interpret/memory.rs

```diff
@@ -844,6 +845,32 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'mir, 'tcx, M> {
         self.get_raw_mut(ptr.alloc_id)?.write_bytes(&tcx, ptr, src)
     }
 
+    /// Writes the given stream of u16s into memory.
+    ///
+    /// Performs appropriate bounds checks.
+    pub fn write_u16s(
+        &mut self,
+        ptr: Scalar<M::PointerTag>,
+        src: impl IntoIterator<Item = u16>,
+    ) -> InterpResult<'tcx> {
+        let src = src.into_iter();
+        assert!(src.size_hint().1.is_some()); // Check that the iterator has an upper bound.
+        let size = Size::from_bytes(u64::try_from(src.size_hint().0).unwrap());
+        let ptr = match self.check_ptr_access(ptr, size, Align::from_bytes(2).unwrap())? {
+            Some(ptr) => ptr,
+            None => return Ok(()), // zero-sized access
+        };
+        let tcx = self.tcx.tcx;
+        let allocation = self.get_raw_mut(ptr.alloc_id)?;
+
+        for (idx, item) in src.enumerate() {
```

Comment:
This fails to check that the `size_hint` did not lie.
I think it is better to loop up to `len` and make sure the iterator did not lie:
```rust
for idx in 0..len {
  let val = Scalar::from_u16(src.next().unwrap());
  ...
}
```